File: 0279-perfect-squares/0279-perfect-squares.java
class Solution {
    public int numSquares(int n) {
        List<Integer> square = new ArrayList<>();

        int x = 1;

        while (x * x <= n){
            square.add(x * x);
            x++;
        }

        int[] dp = new int[n + 1];

        Arrays.fill(dp , Integer.MAX_VALUE);
        dp[0] = 0;

        for(int i = 1; i <= n ; i++){
            for (int sq : square){
                if(i < sq)
                    break;

                dp[i]= Math.min(dp[i] , dp[i - sq] + 1);
            }
        }

        return dp[n];
    }
}

File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        squares = []

        x = 1

        while x * x <= n :
            squares.append(x * x)
            x += 1

        dp = [float('inf')] * (n + 1)
        dp[0] = 0

        for i in range(1 , n + 1):
            for sq in squares :
                if i >= sq :
                    dp[i] = min(dp[i] , dp[i - sq] + 1)

        return dp[n]
        